### PR TITLE
fix(fuse): align link/symlink permission checks with POSIX (#1253)

### LINKED ISSUE #1253: [BUG]: FUSE link and symlink operations ignore POSIX permission boundaries
**Is your feature request related to a problem?**

linkat01 fails with ENOENT where link should succeed; prot_hsymlinks and link01 allow symlink/hardlink creation that should be denied.

linkat01 2 TFAIL linkat failed unexpectedly TEST_ERRNO=ENOENT(2)
link01 1 TFAIL symlink(object, symbolic) was succesful

Reproduction:

1. Mount curvine-fuse
2. Run LTP linkat01
3. Observe ENOENT on valid linkat

**Impact**

Incorrect link semantics can bypass directory permission models.

**Describe the solution you'd like**

linkat should succeed for valid paths; symlink/hardlink across owners should be denied per LTP.

Align link/linkat/symlink permission checks with POSIX.

**Additional context**

Hypothesis pending deeper code trace; see cited issues when available.

Root-cause hypothesis: link/linkat/symlink permission semantics on curvine-fuse are too permissive (hypothesis: missing owner/mode checks).

Evidence summary: linkat01 ENOENT on expected success; prot_hsymlinks and link01 allow forbidden hard/symlink operations.

Related issues: None identified.

Acceptance / test gaps:

- linkat01, prot_hsymlinks, link01 pass

### PR DESCRIPTION
**Pull request by LJF-050**
# Summary

Align Curvine FUSE `link`/`symlink` permission checks with Linux POSIX semantics so directory-entry creation is denied when the parent lacks write/search permission, sticky hard-link ownership rules are enforced, and hard links to dangling symlink inodes succeed.

# Issue Describe / Design

Closes #1253

- **Root cause:** FUSE `link`/`symlink` omitted parent `W_OK|X_OK` checks; sticky `may_link` was not applied; `protected_hardlinks`/`lookup_link` used `fs_stat` on dangling symlink sources (ENOENT); master rejected hard links to `FileType::Link`.
- **Reproduction:** Mount curvine-fuse with permission checks enabled; run LTP `link01` (and related link/symlink cases). Observe symlink succeeding where denial is expected, and/or hard-link-to-dangling-symlink failures.
- **Fix approach:** Enforce parent `W_OK|X_OK` on link/symlink; apply sticky hard-link ownership checks; prefer dcache status and synthesize post-link attrs for dangling symlink targets; allow master hard links to symlink inodes.

# Changes

| Module / File | Change | Impact on existing behavior |
| ------------- | ------ | --------------------------- |
| `curvine-fuse/src/fs/curvine_file_system.rs` | Parent ACL + sticky checks on `link`; dcache-based protected hardlink/status synthesis; parent ACL on `symlink` | Permission denials now match Linux for denied parents; dangling-symlink hard links succeed |
| `curvine-fuse/src/fuse_utils.rs` | Add `check_sticky_hardlink` helper + unit test | None for callers that skip sticky dirs |
| `curvine-fuse/src/fs/dcache/dir_tree.rs` | Unit test that hard links preserve canonical source path | Test-only |
| `curvine-server/src/master/meta/fs_dir.rs` | Allow hard links to `FileType::Link` | Symlink inodes can receive additional hard links |
| `curvine-server/tests/master_fs_test.rs` | Cover hard link to dangling symlink | Test-only |

# Test verified

| Test case | Result | Notes |
| --------- | ------ | ----- |
| `make format` | PASS | Format + clippy gates |
| `cargo test -p curvine-fuse --lib -- check_sticky_hardlink_denies_cross_owner_link hard_link_preserves_source_path` | PASS | |
| `cargo test -p curvine-server --test master_fs_test test_hardlink_to_dangling_symlink_inode` | PASS | |
| LTP originating profile on fixed HEAD | PASS for assigned `link01` | `link01`/`link06`/`linkat01`/`symlink03` TPASS; original failure absent; no new failure identities after semantic attribution of volatile historical cases |

# Dependencies

- Related PRs: None
- Related issues: #1253
- Blocks / blocked by: None

### CODE REVIEW

File: curvine-fuse/src/fs/curvine_file_system.rs [hunk 1]
```diff
@@ -1642,14 +1678,30 @@ impl fs::FileSystem for CurvineFileSystem {
...
 
         self.fs.link(&src_path, &des_path).await?;
-        let attr = self
-            .state
-            .lookup_link(op.header.nodeid, name, oldnodeid)
-            .await?;
+        // After a successful master link, prefer synthesizing the new dentry from the
+        // source inode already in dcache. Fresh fs_stat("nick") can race or miss when
+        // the source was a dangling symlink (LTP link01 case 2).
+        let attr = {
```
Comment (szbr00) [inline curvine-fuse/src/fs/curvine_file_system.rs]: This post-link synthesize path regresses nlink correctness versus the previous always-`lookup_link` flow.

`DirTree::link` does `add_link(1)` then `update_status(status)`, so the status you pass must already carry the final nlink (hence the local `saturating_add(1)`). That means the attr returned to the kernel is `dcache_nlink + 1`, not the master's authoritative count. Problems:
1. Concurrent hard links on the same source can both observe nlink=N and both reply N+1 while master is at N+2.
2. If dcache nlink was already stale, the entry_out nlink stays wrong until a later getattr.
3. There is a TOCTOU window between `dir_read` (clone) and `dir_write` (link); if the inode disappears, master link has already succeeded but this returns an error.

Unless `lookup_link`/`fs_stat(parent, name)` is proven broken for the dangling-symlink case after the master `FileType::Link` fix, prefer always calling `lookup_link` here (master is source of truth for nlink). If synthesize must stay, do the read-modify under one write lock and/or fall back to `lookup_link` on failure, and set nlink from master rather than `dcache + 1`.

File: curvine-fuse/src/fs/curvine_file_system.rs [hunk 2]
```diff
@@ -1710,6 +1762,10 @@ impl fs::FileSystem for CurvineFileSystem {
         let link_path = self.state.get_path_common(id, Some(linkname))?;
         self.ensure_writable_path(&link_path, RpcCode::Symlink)
```
Comment (szbr00) [inline curvine-fuse/src/fs/curvine_file_system.rs]: `link`/`unlink` check parent `W_OK|X_OK` before `ensure_writable_path`; `symlink` does the reverse. Please move this `check_permissions` above `ensure_writable_path` so denial errno prefers EACCES (POSIX parent ACL) over mount read-only errors, consistent with the other directory-entry creation paths.

File: curvine-fuse/src/fs/curvine_file_system.rs (hunk 1 above)
Comment (LJF-050) [reply curvine-fuse/src/fs/curvine_file_system.rs]: Thanks. Fixed in 82958e2: after a successful master link we always call lookup_link so nlink comes from master fs_stat(parent, name), rather than synthesizing dcache_nlink+1. The master FileType::Link hardlink support covers the dangling-symlink case.

File: curvine-fuse/src/fs/curvine_file_system.rs (hunk 2 above)
Comment (LJF-050) [reply curvine-fuse/src/fs/curvine_file_system.rs]: Thanks. Fixed in 82958e2: symlink now checks parent W_OK|X_OK before ensure_writable_path, matching link/unlink so ACL denial prefers EACCES over mount read-only errors.

**Review by szbr00 — APPROVED** [review]
